Docs: Improve documentation of UI extensions (#69364)
grafana/grafana: docs/sources/developers/plugins/extend-the-grafana-ui-with-links.md

---
title: Use extensions to add new functionality in the Grafana user interface
description: Learn how to add links to the Grafana user interface from an app plugin
---

> # The when, why and how of UI extensions

Reviewer: Usually best to start with the Why

---
title: Use extensions to add new functionality in the Grafana user interface
description: Learn how to add links to the Grafana user interface from an app plugin
---

# The when, why and how of UI extensions

Before we go into details we need to cover the two major concepts in the UI extensions feature of Grafana.

> 1. Extension point - a place in the UI where plugins can provide new functionality to the end user. Examples of extension points are the dashboard panel menu or toolbar in explore.

Reviewer: I think we need to make it clearer that Grafana provides some extension points, currently the dashboard panel menu or toolbar in explore, which app plugin developers can hook into; and that App plugins can also provide additional extension points. How about: 

```suggestion
1. Extension point - a place in the UI where plugins can provide new functionality to the end user. The Grafana UI currently exposes extension points in the panel menu within a dashboard, and the toolbar in Explore. App plugins can also define their own extension points.
```

> 2. Extensions - New functionality, usually registered by an app plugin, that will be displayed at an extension point. Examples of extensions are the possibility to create an incident directly from the dashboard panel menu.

Reviewer: nit
```suggestion
2. Extensions - new functionality, usually registered by an app plugin, that will be displayed at an extension point. Examples of extensions are the possibility to create an incident directly from the dashboard panel menu.
```

However, I do think we need to be more explicit on what functionality can be added, or expand the examples in a more generic way i.e. Link/Modal/Cmd etc

Reviewer: i.e. provide a link to navigate a user to a view within a separate plugin, or open a modal menu to allow the user to trigger functionality from a separate plugin within their current context

Author: - You can link the user to a different view within the extending plugin.
- Open a different workflow in a modal that sits on top of the current view.
- Trigger background tasks.

![panel menu](https://user-images.githubusercontent.com/172951/242723354-a10d6238-22f1-4458-b85e-ac3c7f014b22.png)

In the example above we have one extension point with two extensions registered. This highlights one of the benefits of using UI extensions. Once you have added an extension point to your UI it can be extended multiple times by multiple plugins.

> ## When should I add an extension point?

Reviewer: When and Why do blur a little, partly as a result of when coming before Why. I think if we re-order this will naturally clear up. I can help with this.

If you have parts of your UI where it would improve the user experience to allow users to trigger functionality directly from the current view instead of navigating to the plugin. If this is the case, then it might be a good idea to consider adding an extension point.

Examples where it would be useful:

- The user views a dashboard with historical data. By adding an extension point to this part of the UI the Machine Learning app plugin can give the user the ability to create a forecast for that data directly from the panel.
- The user views a firing alert. By adding an extension point to this part of the UI the Incident app plugin can give the user the ability to create an incident directly from the alert view.

> ## Why should I add an extension point?

Reviewer: Missing "Why should I add an extension"

Adding an extension point to your UI gives a number of benefits such as:

- Add the UI extension point once to enable multiple plugins to extend your UI with new functionality. No additional effort needed to provide functionality from more plugins in your UI.
- Clean separation of concerns. Your application does not need to know anything about the plugin extending your UI.
- Integration build for change. Since your application doesn't know anything about the internals of the plugin extending your UI they are free to change their APIs without risk of breaking the extension point UI.

> - Easy to bootstrap. If both apps are installed and enabled the extensions will automatically be configured and displayed to the user.

Reviewer: I think we need to expand on this section. I _believe_ you're referring to the removal of any requirement to query for available extensions or ascertain their availability/health status before deciding on options to present to a user?

Reviewer: Being easy to add generally could be more emphasised, particularly in scenarios where extension points could be requested by those who wish to create an extension.

> - Extensions are fast. We pre build the extensions registry at Grafana boot time which makes it fast to use while rendering the UI.

Reviewer: Does this relate to health checks etc?

Author: Yes, so if we compare it to the "health check"-way of doing things.

## How should I add an extension point to my UI?

> There are a couple of things you need to consider when adding an extension point to your UI. The first one is to define an extension point ID. It is basically just a string describing the part of the UI where the extension point lives. You should be able to figure out where in the UI the extensions will be added by reading the extension point ID.

Reviewer: ```suggestion
There are a couple of things you need to consider when adding an extension point to your UI. The first one is to define an extension point ID. It is basically just a string describing the part of the UI where the extension point lives. A developer looking to register an extension should be able to figure out where in the UI the extensions will be added by reading the extension point ID.
```

> There are a couple of things you need to consider when adding an extension point to your UI. The first one is to define an extension point ID. It is basically just a string describing the part of the UI where the extension point lives. You should be able to figure out where in the UI the extensions will be added by reading the extension point ID.

Reviewer: we should break these consideration points into 'steps'

> Extension points living in core Grafana should start with `grafana/` and extension points living in plugins should start with the plugin id e.g. `myorg-basic-app/`.

Reviewer: ```suggestion
Extension points living in core Grafana start with `grafana/` and extension points living in plugins should start with the plugin id e.g. `myorg-basic-app/`.
```

> The second thing you need to consider is how to design the UI of the extension point so it supports a scenario where multiple extensions are being added without breaking the UI.

Reviewer: Is there more guidance we can provide as examples?

You also need to consider if there are any information from the current view that should be shared with the extensions added to the extension point. It could be information from the current view that could let the extending plugin prefill values etc in the functionality being added via the extension.

Lastly you need to call the `getPluginExtensions` with your extension point ID to receive the list of configured extensions for your extension point.

```typescript
const extensionPointId = 'plugins/myorg-extensionpoint-app/actions';
const context: AppExtensionContext = {
  // Add information that should be shared with the extensions to this object.
};

const { extensions } = getPluginExtensions({
  extensionPointId,
  context,
});
```

# How can I extend a plugin/Grafana with new functionality?

## Before you begin

Be sure your plugin meets the following requirements before proceeding:

- It must be an app plugin.
- It must be preloaded by setting the [preload property]({{< relref "./metadata" >}}) to `true` in the `plugin.json`.
- It must be installed and enabled.

> ## Available extension points within Grafana

Reviewer: This now duplicates content
